KEP-3962: MutatingAdmissionPolicy Alpha

jpbetz opened the pull request:
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

Implements alpha of https://github.com/kubernetes/enhancements/issues/3962

This supersedes https://github.com/kubernetes/kubernetes/pull/126497 to include:

- [x] API Updated to support JSON Patch and Apply Configuration mutations
- [x] Remove message and messageExpression since they have no obvious relevance for mutation
- [x] JSON Patch implementation + "jsonpatch.escapeKey()" CEL function
- [x] Variables
- [x] Runtime cost
- [x] request variable
- [x] authorizer variable
- [x] type check: CEL Object types must match field name paths
- [x] CRD version conversion
- [x] matchConditions
- [x] failurePolicy handling
- [x] relocation of reinvocationPolicy field to spec level
- [x] Disallow mutation of atomics with apply configurations
- [x] integration testing of: CRDs, CRD version conversion, matchConditions, failurePolicy, subresources, authorizers, request variables, json patch

#### Which issue(s) this PR fixes:


#### Special notes for your reviewer:

- CEL expressions for Validating and Mutating admission policy are not type checked at compile time because there is no guarantee that the type to be validated exists yet (for CRDs), and the expressions are allowed to match multiple types. This has lots of ripple effects in this PR.  At runtime we do some additional type checking.. I've added some type mismatch checking to avoid obvious errors, but there are still significant gaps, even at runtime. I have additional work not yet in this PR that takes the type checking further, but I figured it would be good to get a baseline implementation reviewed and then iterate.

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Introduce v1alpha1 API for mutating admission policies, enabling extensible admission control via CEL expressions (KEP  3962: Mutating Admission Policies). To use, enable the `MutatingAdmissionPolicy` feature gate and the `admissionregistration.k8s.io/v1alpha1` API via `--runtime-config`.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

Please use the following format for linking documentation:

- [KEP](https://github.com/kubernetes/enhancements/pull/4399)

```docs

```

jpbetz commented:
/label api-review

This needs additional tests but is otherwise feature complete and ready for API review.

jpbetz commented:
/triage accepted
/priority important-longterm

deads2k commented on pkg/apis/admissionregistration/validation/validation.go:
```diff
@@ -1328,3 +1382,243 @@ func isCELIdentifier(name string) bool {
+			accessor := &mutating.JSONPatchCondition{
+				ValueExpression: trimmedValueExpression,
+			}
+			result := compiler.CompileCELExpression(accessor, plugincel.OptionalVariableDeclarations{HasParams: paramKind != nil, HasAuthorizer: true, StrictCost: true}, envType)
```
Do we have any way to try to guess whether we'll get a valid yaml object after evaluating the expression?  Even if we cannot check that types are correct, can we take a guess at syntactic validity?  Even a warning if we can get it mostly right could help.

jpbetz replied:
Good question.  All CEL data literal declarations has maps, arrays and json compatible scalar types.  But we do have a few special types floating around like authorizer...   I'll see if there is anything we can do in the type checker to detect these.

jpbetz replied:
I looked into this.  There's not a lot we more we can do at compile time.  There are a few constraints here.  We don't always know what resource is being mutated statically, and we don't know the type of what the json patch path refers to, so we can't do any type matching at those levels.  Because of that, we're telling the CEL compiler that the return type is `cel.DynType`  and letting the expression return most anything.  We really want a subset of DynType that's just types that map to JSON values, but there's nothing in the type system available to achieve that.  The good news is that MOST types do map to JSON values, so there is a relatively small set of special types that will fail at runtime, but I don't think we can detect them.

deads2k commented on staging/src/k8s.io/api/admissionregistration/v1alpha1/types.go:
```diff
@@ -663,3 +663,346 @@ const (
+	// If this is unset, all resources matched by the policy are validated by this binding
+	// When resourceRules is unset, it does not constrain resource matching. If a resource is matched by the other fields of this object, it will be validated.
+	// Note that this is differs from MutatingAdmissionPolicy matchConstraints, where resourceRules are required.
+	// Only the CREATE and UPDATE operations are allowed.  DELETE and CONNECT operations may not be matched.
```
Why disallow CONNECT?  That normally passes the options through.

disallowing delete I understand since it always passes a nil object, so any control on delete should be done via validating.

jpbetz replied:
> That normally passes the options through.

Ah, I didn't realize. I'll add it back.

deads2k reviewed (changes requested):
Reviewed the API, defaulting, and validation.  Some comments.

jpbetz commented on staging/src/k8s.io/api/admissionregistration/v1alpha1/types.go:
```diff
@@ -663,3 +663,346 @@ const (
+	// path specifies the path to the target value.
+	// path must be a '/' delimited field path conforming to the [JSON Path spec](https://jsonpatch.com/).
+	// Required.
+	Path string `json:"path" protobuf:"bytes,2,opt,name=path"`
```
I think this might be needed.

The digs into the bigger topic of json patch expressiveness, so  I should probably at least summarize two directional alternatives:

Option 1: Define the JSON Patch in the k8s API, with expression support for some fields (what I have in this PR now)

```
jsonPatch:
  - op: "add"
    path: "/metadat/label/x"
    valueExpression: "'y'"
```

Option 2: Have CEL return the entire JSON Patch:

```
jsonPatch:
  expression: `
     JSONPatch{
       [
         Op: "add",
         Path: "/metadata/labels/x",
         Value: "y"
       ]
     }
  `
````

If we continue with Option 1, we can add `patchExpression` and `fromExpression` to get some addtional expressiveness.

If we go with Option 2, it becomes possible to do things like use JSON Patch to modify all containers in a pod, which will never be possible with Option 1.

I'd admit I started building out Option 1 without seriously considering such use cases.  But they're on my mind.

ApplyConfigurations can update all container in a pod quite easily, but have limitations around removing values and updating atomic maps/lists/structs.

jpbetz replied:
I've switched `path` and `from` to CEL expressions (`pathExpression` and `fromExpression`).  Happy to discuss expressiveness topic further.

jpbetz commented on pkg/apis/admissionregistration/v1alpha1/defaults.go:
```diff
@@ -57,3 +58,19 @@ func SetDefaults_ParamRef(obj *admissionregistrationv1alpha1.ParamRef) {
 	}
 }
+
+// SetDefaults_MutatingAdmissionPolicySpec sets defaults for MutatingAdmissionPolicySpec
+func SetDefaults_MutatingAdmissionPolicySpec(obj *admissionregistrationv1alpha1.MutatingAdmissionPolicySpec) {
+	if obj.FailurePolicy == nil {
+		policy := admissionregistrationv1alpha1.Fail
+		obj.FailurePolicy = &policy
+	}
+}
+
+// SetDefaults_Mutation sets defaults for Mutation
+func SetDefaults_Mutation(obj *admissionregistrationv1alpha1.Mutation) {
+	if obj.ReinvocationPolicy == nil {
+		never := admissionregistrationv1.NeverReinvocationPolicy
+		obj.ReinvocationPolicy = &never
+	}
+}
```
Looks like it was done this way because mutating webhooks defined reinvocation policy this way.

That said, reinvocationPolicy was added to webhooks after the API was already established.  I'm OK with changing it if there are no objections to having the field behave differently than how it does on mutating webhooks.

jpbetz replied:
Switched to a required field here.

jpbetz commented on pkg/apis/admissionregistration/validation/validation.go:
```diff
@@ -597,6 +640,17 @@ func ignoreValidatingAdmissionPolicyMatchConditions(new, old *admissionregistrat
+	if !reflect.DeepEqual(new.Spec.ParamKind, old.Spec.ParamKind) {
+		return false
+	}
+	if !reflect.DeepEqual(new.Spec.MatchConditions, old.Spec.MatchConditions) {
```
Fixed. I'll change this for ValidatingAdmissionPolicy too, which looks safe since it is used for ratcheting and semantic equality widens which expressions can be ratcheted.

jpbetz commented on pkg/apis/admissionregistration/validation/validation.go:
```diff
@@ -1328,3 +1382,243 @@ func isCELIdentifier(name string) bool {
+		allErrors = append(allErrors, field.Required(fldPath.Child("mutations"), "mutations must contain at least one item"))
+	} else {
+		for i, mutation := range spec.Mutations {
+			allErrors = append(allErrors, validateMutation(getCompiler(), &mutation, spec.ParamKind, opts, fldPath.Child("mutations").Index(i))...)
```
`getCompiler()` is copy-pasted from ValidatingAdmissionPolicy's validation. It is a "lazy getOrCreate" function, so there is only one compiler for all mutations...   but I don't see any value here since we always validate at least one CEL expression. I'll simplify.

jpbetz commented on staging/src/k8s.io/api/admissionregistration/v1alpha1/types.go:
```diff
@@ -663,3 +663,346 @@ const (
+	// mutations that specify this option *must* be idempotent, able to process objects they previously admitted.
+	// Note:
+	// * the number of additional invocations is not guaranteed to be exactly one.
+	// * if additional invocations result in further modifications to the object, webhooks are not guaranteed to be invoked again.
```
Looks like this got copied from webhooks.  I'll rephrase.

jpbetz replied:
I remember that with webhooks we didn't want to code-ify into the API that reinvocation was limited to one retry.  Are we still OK with that or do we want to commit to "only one reinvocation" here in the API?

jpbetz commented:
/retest

jpbetz commented:
/assign @sttts
For implementation review.

jpbetz commented:
/retest

jpbetz commented:
@deads2k I've changed JSON Patch to be declared entirely in CEL and moved reinvocationPolicy up to the spec level.  This is ready for another API review pass.

jpbetz commented:
/retest
CI got starved

deads2k commented on pkg/apis/admissionregistration/types.go:
```diff
@@ -1163,3 +1164,317 @@ type MatchCondition struct {
+// +k8s:prerelease-lifecycle-gen:introduced=1.32
+
+// MutatingAdmissionPolicy describes an admission policy that may mutate an object.
+type MutatingAdmissionPolicy struct {
```
Since we have parameters, don't we need the status to represent errors that are not resolvable until the parameter types are known?

jpbetz replied:
OK if we add this before beta but not in 1.32? I have a branch that can do the compilation correctly but it's more complicated than VAP due to the new types introduced.

deads2k commented on staging/src/k8s.io/api/admissionregistration/v1alpha1/types.go:
```diff
@@ -663,3 +663,333 @@ const (
+	// The AdmissionPolicy cares about a request if it matches _all_ Constraints.
+	// However, in order to prevent clusters from being put into an unstable state that cannot be recovered from via the API
+	// MutatingAdmissionPolicy cannot match MutatingAdmissionPolicy and MutatingAdmissionPolicyBinding.
+	// Only the CREATE and UPDATE operations are allowed.  DELETE and CONNECT operations may not be matched.
```
old comment about connect?  We decided to match regular admission, right?

jpbetz replied:
Yes, good catch. I'll fix this doc string.

deads2k commented on staging/src/k8s.io/api/admissionregistration/v1alpha1/types.go:
```diff
@@ -663,3 +663,333 @@ const (
+	//
+	// CEL expressions have access to the contents of the API request/response, organized into CEL variables as well as some other useful variables:
+	//
+	// - 'object' - The object from the incoming request. The value is null for DELETE requests.
+	// - 'oldObject' - The existing object. The value is null for CREATE requests.
+	// - 'request' - Attributes of the API request([ref](/pkg/apis/admission/types.go#AdmissionRequest)).
+	// - 'params' - Parameter resource referred to by the policy binding being evaluated. Only populated if the policy has a ParamKind.
+	// - 'namespaceObject' - The namespace object that the incoming object belongs to. The value is null for cluster-scoped resources.
+	// - 'variables' - Map of composited variables, from its name to its lazily evaluated value.
```
VAP can use the authorizer to make decisions about allowed or disallowed.  I expected the capability here to be peer to that.

jpbetz replied:
Oh yes. I've included the authorizer in the implementation, so that it's capabilities can be used when patching. I'll add documentation about it.

jpbetz replied:
Added docs. Implementation has support.

deads2k commented on staging/src/k8s.io/api/admissionregistration/v1alpha1/types.go:
```diff
@@ -663,3 +663,333 @@ const (
+	//
+	// CEL expressions have access to the contents of the API request/response, organized into CEL variables as well as some other useful variables:
+	//
+	// - 'object' - The object from the incoming request. The value is null for DELETE requests.
+	// - 'oldObject' - The existing object. The value is null for CREATE requests.
+	// - 'request' - Attributes of the API request([ref](/pkg/apis/admission/types.go#AdmissionRequest)).
+	// - 'params' - Parameter resource referred to by the policy binding being evaluated. Only populated if the policy has a ParamKind.
+	// - 'namespaceObject' - The namespace object that the incoming object belongs to. The value is null for cluster-scoped resources.
+	// - 'variables' - Map of composited variables, from its name to its lazily evaluated value.
+	//   For example, a variable named 'foo' can be accessed as 'variables.foo'.
```
VAP can use the authorizer to make decisions about allowed or disallowed.  I expected the capability here to be peer to that.

jpbetz replied:
Added docs. Implementation has support.

deads2k commented on pkg/apis/admissionregistration/validation/validation.go:
```diff
@@ -495,6 +517,18 @@ var supportedValidationPolicyReason = sets.NewString(
+	string(admissionregistration.PatchTypeJSONPatch),
+)
+
+// MutatatingAdmissionPolicy does not support DELETE or CONNECT
```
allowing connect as standard admission?

jpbetz replied:
Fixed.

jpbetz commented:
Thanks @deads2k. Feedback applied.

deads2k commented on staging/src/k8s.io/api/admissionregistration/v1alpha1/types.go:
```diff
@@ -663,3 +663,341 @@ const (
+	// The AdmissionPolicy cares about a request if it matches _all_ Constraints.
+	// However, in order to prevent clusters from being put into an unstable state that cannot be recovered from via the API
+	// MutatingAdmissionPolicy cannot match MutatingAdmissionPolicy and MutatingAdmissionPolicyBinding.
+	// Only the CREATE and UPDATE operations are allowed.  DELETE and CONNECT operations may not be matched.
+	// '*' matches only CREATE and UPDATE.
```
```suggestion
	// Only the CREATE, UPDATE, and CONNECT operations are allowed.
	// '*' matches only CREATE, UPDATE, and CONNECT.
```

is this correct?  I think that is wha tis needed for parity

jpbetz replied:
Oops, yes. I missed this reference.

deads2k reviewed (changes requested):
the API lgtm otherwise

jpbetz commented:
/retest

sttts commented on staging/src/k8s.io/api/admissionregistration/v1alpha1/types.go:
```diff
@@ -663,3 +663,341 @@ const (
+	// occur from CEL expression parse errors, type check errors, runtime errors and invalid
+	// or mis-configured policy definitions or bindings.
+	//
+	// A policy is invalid if spec.paramKind refers to a non-existent Kind.
```
nit: no need to qualify paramKind here

sttts commented on staging/src/k8s.io/api/admissionregistration/v1alpha1/types.go:
```diff
@@ -663,3 +663,341 @@ const (
+	FailurePolicy *FailurePolicyType `json:"failurePolicy,omitempty" protobuf:"bytes,5,opt,name=failurePolicy,casttype=FailurePolicyType"`
+
+	// matchConditions is a list of conditions that must be met for a request to be validated.
+	// Match conditions filter requests that have already been matched by the rules,
+	// namespaceSelector, and objectSelector. An empty list of matchConditions matches all requests.
```
Nit:
```suggestion
	// Match conditions filter requests that have already been matched by the matchConstraints.
	// An empty list of matchConditions matches all requests.
```

This would make it more precise and findable.

sttts commented on staging/src/k8s.io/api/admissionregistration/v1alpha1/types.go:
```diff
@@ -663,3 +663,341 @@ const (
+	// expression will be evaluated by CEL to create an apply configuration.
+	// ref: https://github.com/google/cel-spec
+	//
+	// Apply configurations are declared in CEL using object initialization.  For example, this CEL expression
```
```suggestion
	// Apply configurations are declared in CEL using object initialization. For example, this CEL expression
```

sttts commented on staging/src/k8s.io/api/admissionregistration/v1alpha1/types.go:
```diff
@@ -663,3 +663,341 @@ const (
+	// - 'Object.<fieldName>' - CEL type of object field (such as 'Object.spec')
+	// - 'Object.<fieldName1>.<fieldName2>...<fieldNameN>` - CEL type of nested field (such as 'Object.spec.containers')
+	//
+	// CEL expressions have access to the contents of the API request/response, organized into CEL variables as well as some other useful variables:
```
of the response?

jpbetz replied:
Good catch. Not sure why this is here. I'll remove and backtrack to see where else we use this phrasing.

sttts commented on staging/src/k8s.io/api/admissionregistration/v1alpha1/types.go:
```diff
@@ -663,3 +663,341 @@ const (
+	//
+	//	  [
+	//      JSONPatch{op: "test", path: "/spec/example", value: "Red"},
+	//	    JSONPatch{op: "replace", path: "/spec/example", value: "Green"}
```
indention wrong?

sttts commented on staging/src/k8s.io/api/admissionregistration/v1alpha1/types.go:
```diff
@@ -663,3 +663,341 @@ const (
+	//	    JSONPatch{
+	//	      op: "add",
+	//	      path: "/spec/selector",
+	//	      value: Object.spec.selector{matchLabels: {"environment": "test"}}
```
Below you write "strings, integers, arrays, maps and objects". Does it mean either can be used?

Are objects type-checked to match the given JSON path?

jpbetz replied:
> Below you write "strings, integers, arrays, maps and objects". Does it mean either can be used?

Yeah, any of these can be used.

> Are objects type-checked to match the given JSON path?

We can't because the policy might be matching a kind that does not exist yet, or it might matching multiple types or "*".  We will add "informational type checking" like VAP supports before going to beta, we can include this in what we type check for that.

sttts commented on staging/src/k8s.io/api/admissionregistration/v1alpha1/types.go:
```diff
@@ -663,3 +663,341 @@ const (
+	// - 'authorizer.requestResource' - A CEL ResourceCheck constructed from the 'authorizer' and configured with the
+	//   request resource.
+	//
+	// The `apiVersion`, `kind`, `metadata.name` and `metadata.generateName` are always accessible from the root of the
```
what does "always" mean?

sttts commented on staging/src/k8s.io/api/admissionregistration/v1alpha1/types.go:
```diff
@@ -663,3 +663,341 @@ const (
+	//   request resource.
+	//
+	// The `apiVersion`, `kind`, `metadata.name` and `metadata.generateName` are always accessible from the root of the
+	// object. No other metadata properties are accessible.
```
What's the best place to read about the why of this restriction? The lack of namespace, labels and annotations is surprising here.

jpbetz replied:
This is inaccurate. CRD Validation rules have this restriction, and this text got copied over from there to VAP and then to here. Removing.

sttts commented on staging/src/k8s.io/api/admissionregistration/v1alpha1/types.go:
```diff
@@ -663,3 +663,341 @@ const (
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +k8s:prerelease-lifecycle-gen:introduced=1.32
+
+// MutatingAdmissionPolicyBinding binds the MutatingAdmissionPolicy with paramerized resources.
```
```suggestion
// MutatingAdmissionPolicyBinding binds the MutatingAdmissionPolicy with parametrized resources.
```

sttts commented on staging/src/k8s.io/api/admissionregistration/v1alpha1/types.go:
```diff
@@ -663,3 +663,341 @@ const (
+// +k8s:prerelease-lifecycle-gen:introduced=1.32
+
+// MutatingAdmissionPolicyBinding binds the MutatingAdmissionPolicy with paramerized resources.
+// MutatingAdmissionPolicyBinding and parameter CRDs together define how cluster administrators configure policies for clusters.
```
does it have to be a CRD? Couldn't I bind with a configmap?

jpbetz replied:
Good catch. It can be any time.

jpbetz replied:
Clarified.

sttts commented on staging/src/k8s.io/api/admissionregistration/v1alpha1/types.go:
```diff
@@ -663,3 +663,341 @@ const (
+// params, otherwise N is the number of parameters selected by the binding.
+//
+// The CEL expressions of a policy must have a computed CEL cost below the maximum
+// CEL budget. Each evaluation of the policy is given an independent CEL cost budget.
```
what is the CEL budget? Can we reference something in the docs?

sttts commented on staging/src/k8s.io/api/admissionregistration/v1alpha1/types.go:
```diff
@@ -663,3 +663,341 @@ const (
+	PolicyName string `json:"policyName,omitempty" protobuf:"bytes,1,rep,name=policyName"`
+
+	// paramRef specifies the parameter resource used to configure the admission control policy.
+	// It should point to a resource of the type specified in ParamKind of the bound MutatingAdmissionPolicy.
```
```suggestion
	// It should point to a resource of the type specified in spec.ParamKind of the bound MutatingAdmissionPolicy.
```

sttts commented on staging/src/k8s.io/api/admissionregistration/v1alpha1/types.go:
```diff
@@ -663,3 +663,341 @@ const (
+	// MatchResources declares what resources match this binding and will be validated by it.
+	// Note that this is intersected with the policy's matchConstraints, so only requests that are matched by the policy can be selected by this.
+	// If this is unset, all resources matched by the policy are validated by this binding
+	// When resourceRules is unset, it does not constrain resource matching. If a resource is matched by the other fields of this object, it will be validated.
```
With`excludeResourceRules` to be also a thing this reads very vague.

`resourceRules` says:
```
// The policy cares about an operation if it matches _any_ Rule.
```
It's not clear to me whether in `ValidatingAdmissionPolicies` and `MutatingAdmissionPolicies` the "matches _any_ Rule" means that not specifying `resourceRules` means everything or nothing. And with that, it is unclear to me whether here in the binding "it does not constrain resource matching" is redundant or says that we differ from the policy semantics of the same field.

jpbetz replied:
Good catch. The wording does not  account for the exclude rules. I'll rephrase.

sttts commented on pkg/generated/openapi/openapi_test.go:
```diff
@@ -18,10 +18,12 @@ package openapi
-	"reflect"
+	"github.com/go-openapi/jsonreference"
+	"github.com/google/go-cmp/cmp/cmpopts"
 	"testing"
```
order

sttts commented on pkg/generated/openapi/openapi_test.go:
```diff
@@ -51,8 +53,16 @@ func TestOpenAPIRoundtrip(t *testing.T) {
+			opts := []cmp.Option{
+				cmpopts.EquateEmpty(),
+				// jsonreference.Ref contains unexported fields. Compare
+				// by string representation provides a consistent
```
the switch to go cmp I mean.

jpbetz replied:
Something about this API change caused our generated OpenAPI to contain `jsonreference.Ref` entries.  jsonreference.Ref contains unexported fields and caused this test to fail.  I'm a bit surprised I was the first to hit this TBH, since the generator has had the ability generate output like this all along, but this fixes the equality check in a sane/targeted way.

sttts commented on pkg/kubeapiserver/options/plugins.go:
```diff
@@ -160,6 +163,7 @@ func DefaultOffAdmissionPlugins() sets.Set[string] {
 		defaultingressclass.PluginName,          // DefaultIngressClass
 		podsecurity.PluginName,                  // PodSecurity
 		validatingadmissionpolicy.PluginName,    // ValidatingAdmissionPolicy, only active when feature gate ValidatingAdmissionPolicy is enabled
+		mutatingadmissionpolicy.PluginName,      // Mutatingadmissionpolicy, only active when feature gate MutatingAdmissionpolicy is enabled
```
nit: any reasons not to order always `mutating, validating` or the inverse?

jpbetz replied:
Also, not sure the original thinking here.  But I favor of always ordering as `mutating`, `validating`, so I'll fix this occurance.

sttts commented on pkg/kubeapiserver/options/plugins.go:
```diff
@@ -160,6 +163,7 @@ func DefaultOffAdmissionPlugins() sets.Set[string] {
 		defaultingressclass.PluginName,          // DefaultIngressClass
 		podsecurity.PluginName,                  // PodSecurity
 		validatingadmissionpolicy.PluginName,    // ValidatingAdmissionPolicy, only active when feature gate ValidatingAdmissionPolicy is enabled
```
this is GA now, right? So we can drop the comment.

jpbetz replied:
I'll leave it in a bit longer now that we have compatibility version merged and are tracking feature gates post-GA.

sttts commented on pkg/printers/internalversion/printers_test.go:
```diff
@@ -7031,6 +7031,18 @@ func TestTableRowDeepCopyShouldNotPanic(t *testing.T) {
+			},
+		},
+		{
+			name: "MutaatingAdmissionPolicyBinding",
```
```suggestion
			name: "MutatingAdmissionPolicyBinding",
```

jpbetz replied:
Good catch.

sttts commented on pkg/registry/admissionregistration/mutatingadmissionpolicy/authz.go:
```diff
@@ -0,0 +1,105 @@
+/*
+Copyright 2022 The Kubernetes Authors.
```
2024

sttts commented on pkg/registry/admissionregistration/mutatingadmissionpolicy/authz.go:
```diff
@@ -0,0 +1,105 @@
+	}
+
+	// for superuser, skip all checks
+	if rbacregistry.EscalationAllowed(ctx) {
```
This is a little misuse of the method (preexisting in the validating policy case). Wouldn't mind to make it explicit: `isSystemAdmin` or similar.

jpbetz replied:
OK if I follow up on this separately since we'll be changing ValidatingAdmissionPolicy as well?  I've opened https://github.com/kubernetes/kubernetes/issues/127935 to track.

sttts commented on pkg/registry/admissionregistration/mutatingadmissionpolicy/doc.go:
```diff
@@ -1,5 +1,5 @@
 /*
```
wild interpretation of Github that this is a file move :D

sttts commented on pkg/registry/admissionregistration/mutatingadmissionpolicybinding/authz.go:
```diff
@@ -0,0 +1,117 @@
+		return err
+	}
+	if d != authorizer.DecisionAllow {
+		return fmt.Errorf(`user %v does not have "%v" permission on the object referenced by paramRef`, verb, user)
```
with all the code paths above with missing policy, failed REST mapping, it's not super clear what was it that made this authorization fail. Should we be a little bit more verbose which permission we are actually checking? I.e. group, version, resource.

jpbetz replied:
I agree. It would be hard to know what to do as a user given only this error message.

I've expanded this to return different errors for these cases. I added tests to cover each.

sttts commented on pkg/registry/admissionregistration/mutatingadmissionpolicybinding/authz_test.go:
```diff
@@ -0,0 +1,162 @@
+					Resource: "params",
+				}, nil
+			},
+			expectErr: true,
```
case with NotFound on policy get is missing.

sttts replied:
cases missing when relevant fields are NOT updated

jpbetz replied:
I've added a 'param not found' case.  I'm not clear on what a case for relevant fields NOT updated would entail. Can you say any more about what you noticed missing there?

jpbetz replied:
I added a few other tests as well. PTAL

sttts replied:
I meant calling `errs := strategy.ValidateUpdate(ctx, obj, obj)` with 
```
auth: func(ctx context.Context, a authorizer.Attributes) (authorized authorizer.Decision, reason string, err error) {
	return authorizer.DecisionDeny, "", nil
},
```
Then `errs` should be empty.

jpbetz replied:
Thanks, added.

jpbetz commented on staging/src/k8s.io/api/admissionregistration/v1alpha1/types.go:
```diff
@@ -663,3 +663,341 @@ const (
+	// However, in order to prevent clusters from being put into an unstable state that cannot be recovered from via the API
+	// MutatingAdmissionPolicy cannot match MutatingAdmissionPolicy and MutatingAdmissionPolicyBinding.
+	// Only the CREATE, UPDATE, and CONNECT operations are allowed.
+	// '*' matches only CREATE, UPDATE, and CONNECT.
```
I can simplify this to just "The CREATE, UPDATE, and CONNECT operations are allowed."

jpbetz commented on pkg/kubeapiserver/default_storage_factory_builder.go:
```diff
@@ -75,6 +76,8 @@ func NewStorageFactoryConfig() *StorageFactoryConfig {
 		networking.Resource("ipaddresses").WithVersion("v1beta1"),
 		networking.Resource("servicecidrs").WithVersion("v1beta1"),
+		admissionregistration.Resource("mutatingadmissionpolicies").WithVersion("v1alpha1"),
+		admissionregistration.Resource("mutatingadmissionpolicybindings").WithVersion("v1alpha1"),
```
No idea why this was put here. Removing.

jpbetz replied:
Tuns out I had to put this back in. With it removed, `make update` failed with:

```
Running update-openapi-spec
!!! [1008 12:23:59] Timed out waiting for apiserver:  to answer at https://127.0.0.1:8050/healthz; tried 30 waiting 1 between each
!!! [1008 12:23:59] Here are the last 10 lines from kube-apiserver (/tmp/update-openapi-spec.sh.p7CV8g/openapi-api-server.log)
!!! [1008 12:23:59] === BEGIN OF LOG ===
I1008 12:23:32.198635 3378974 store.go:1641] "Monitoring resource count at path" resource="validatingadmissionpolicybindings.admissionregistration.k8s.io" path="<storage-prefix>//validatingadmissionpolicybindings"
I1008 12:23:32.199345 3378974 cacher.go:463] cacher (validatingadmissionpolicybindings.admissionregistration.k8s.io): initialized
I1008 12:23:32.199383 3378974 reflector.go:368] Caches populated for *admissionregistration.ValidatingAdmissionPolicyBinding from storage/cacher.go:/validatingadmissionpolicybindings
I1008 12:23:32.200083 3378974 store.go:1641] "Monitoring resource count at path" resource="validatingadmissionpolicies.admissionregistration.k8s.io" path="<storage-prefix>//validatingadmissionpolicies"
I1008 12:23:32.200834 3378974 cacher.go:463] cacher (validatingadmissionpolicies.admissionregistration.k8s.io): initialized
I1008 12:23:32.200873 3378974 reflector.go:368] Caches populated for *admissionregistration.ValidatingAdmissionPolicy from storage/cacher.go:/validatingadmissionpolicies
I1008 12:23:32.201386 3378974 store.go:1641] "Monitoring resource count at path" resource="validatingadmissionpolicybindings.admissionregistration.k8s.io" path="<storage-prefix>//validatingadmissionpolicybindings"
I1008 12:23:32.202185 3378974 cacher.go:463] cacher (validatingadmissionpolicybindings.admissionregistration.k8s.io): initialized
I1008 12:23:32.202235 3378974 reflector.go:368] Caches populated for *admissionregistration.ValidatingAdmissionPolicyBinding from storage/cacher.go:/validatingadmissionpolicybindings
E1008 12:23:32.202872 3378974 run.go:72] "command failed" err="problem initializing API group \"admissionregistration.k8s.io\": storage codec doesn't seem to match given type: internal type not encodable: no kind \"MutatingAdmissionPolicy\" is registered for version \"admissionregistration.k8s.io/v1\" in scheme \"pkg/api/legacyscheme/scheme.go:30\""
```

jpbetz commented on pkg/registry/admissionregistration/mutatingadmissionpolicybinding/authz.go:
```diff
@@ -0,0 +1,117 @@
+}
+
+func (v *mutatingAdmissionPolicyBindingStrategy) authorize(ctx context.Context, binding *admissionregistration.MutatingAdmissionPolicyBinding) error {
+	if v.authorizer == nil || v.resourceResolver == nil || binding.Spec.ParamRef == nil {
```
Fixed in latest to be hard-required.

jpbetz commented on pkg/registry/admissionregistration/mutatingadmissionpolicy/authz_test.go:
```diff
@@ -0,0 +1,110 @@
+				obj := validMutatingAdmissionPolicy()
+				objWithUpdatedParamKind := obj.DeepCopy()
+				objWithUpdatedParamKind.Spec.ParamKind.APIVersion += "1"
+				errs := strategy.ValidateUpdate(ctx, obj, objWithUpdatedParamKind)
```
Can you say any more about what you mean by 'non-mutating'?

jpbetz replied:
I looked around and couldn't find the spot where we ignore validation if the mutation was a noop. Sorry if it's obvious.  If you point me to it I'll add a test.

sftim commented on pkg/kubeapiserver/options/plugins_test.go:
```diff
@@ -22,9 +22,9 @@ import (
 
 func TestAdmissionPluginOrder(t *testing.T) {
-	// Ensure the last four admission plugins listed are webhooks, quota, and deny
+	// Ensure the last four admission plugins listed are policies, webhooks, quota, and deny
```
This doesn't look right. The order is correct, but what we say test no longer holds. We're now checking for mutating before validating and, within that, policy before external callout(s).

jpbetz replied:
Good point. I've rephrased.

sftim commented on test/integration/apiserver/admissionwebhook/admission_test.go:
Maybe it's time to rename this file?

sftim reviewed (commented):
I went looking to see how we'd done the end to end testing for this, and I didn't see any. If that's because it's not done yet: is it useful to have some for alpha?

jpbetz commented:
> I went looking to see how we'd done the end to end testing for this, and I didn't see any. If that's because it's not done yet: is it useful to have some for alpha?

It's mostly because this feature does not require any nodes or contollers to fully test, so we were able to write the tests against an apiserver started in an integration test.  We will need to add e2e tests eventually, primary for conformance, but that's not super urgent, the "end to end" coverage we get with the integration tests is really solid.

sftim commented on test/integration/apiserver/admissionwebhook/admission_test.go:
Maybe time to rename this file too (if possible)?

jpbetz replied:
I think that makes sense. Purely to minimize rebase problems, I'd like to separate such a change from this PR.

alexzielenski commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/policy/mutating/patch/smd.go:
```diff
@@ -0,0 +1,193 @@
+		return nil, fmt.Errorf("failed to merge patch: %w", err)
+	}
+
+	// Our mutating admission policy sets the fields but does not track ownership.
+	// Newly introduced fields in the patch won't be tracked by a field manager
+	// (so if the original object is updated again but the mutating policy is
+	// not active, the fields will be dropped).
+	//
+	// This necessarily means that changes to an object by a mutating policy
+	// are only preserved if the policy was active at the time of the change.
+	// (If the policy is not active, the changes may be dropped.)
```
I never verified the veracity of this comment

jpbetz replied:
I'll follow up. This does feel like something that applies generally to all mutations, so I'm not super concerned from the perspective of this particular alpha enhancement, but it would be good to have this gotcha well understood and codified into our docs.

alexzielenski commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/policy/mutating/plugin.go:
```diff
@@ -0,0 +1,151 @@
+			)
+		},
+		func(a authorizer.Authorizer, m *matching.Matcher, client kubernetes.Interface) generic.Dispatcher[PolicyHook] {
+			return NewDispatcher(a, m, patch.NewTypeConverterManager(nil, client.Discovery().OpenAPIV3()))
```
This is passing `nil` for the static type converter meaning all schemas are being parsed through openapi.  good optimization for the future is to complete this wiring

jpbetz replied:
+100, I'll add a TODO in the code.

sttts commented on pkg/registry/admissionregistration/mutatingadmissionpolicy/doc.go:
```diff
@@ -1,5 +1,5 @@
 /*
-Copyright 2024 The Kubernetes Authors.
+Copyright 2022 The Kubernetes Authors.
```
No strong opinion. Thought it was copied from the validating case. In that case I would expect 2024.

sttts commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/authorizer/caching_authorizer.go:
```diff
@@ -39,7 +39,10 @@ type cachingAuthorizer struct {
 }
 
-func newCachingAuthorizer(in authorizer.Authorizer) authorizer.Authorizer {
+// NewCachingAuthorizer returns and author that caches decisions for the duration
```
```suggestion
// NewCachingAuthorizer returns an authorizer that caches decisions for the duration
```

sttts commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/authorizer/caching_authorizer.go:
```diff
@@ -39,7 +39,10 @@ type cachingAuthorizer struct {
-func newCachingAuthorizer(in authorizer.Authorizer) authorizer.Authorizer {
+// NewCachingAuthorizer returns and author that caches decisions for the duration
+// of the authorizers use.  Intended to be used for short-lived operations such as
+// the handling of a request in the admission chain, and then discarded.
```
am curious, what is the reason we need this on-top of a "normal" caching authorizer?

jpbetz replied:
This is the authorizer added by https://github.com/kubernetes/kubernetes/pull/116443.  I'm not familiar with how it compares with the "normal" one, but it's targeted specifically at caching only for the scope of a single request's admission chain handling, and is intended to cache results for identical admission checks performed from CEL.

sttts commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/policy/generic/interfaces.go:
```diff
@@ -49,6 +49,9 @@ type Source[H Hook] interface {
 // active hooks returned by the source.
 type Dispatcher[H Hook] interface {
+	// Run the dispatcher. This method should be called only once at startup.
+	Run(ctx context.Context) error
```
it blocks? In most case in the codebase Run=blocks, Start=returns

jpbetz replied:
Looks like it starts goroutines and routines.

@alexzielenski does this match your understanding?  Should we rename?

jpbetz replied:
Great, thanks for confirming. I'll make the edit

sttts commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/policy/generic/policy_dispatcher.go:
```diff
@@ -148,17 +151,16 @@ func (d *policyDispatcher[P, B, E]) Dispatch(ctx context.Context, a admission.At
 			} else if !matches {
 				continue
+			} else if _, err = versionedAttrAccessor.VersionedAttribute(matchGVK); err != nil {
+				// VersionedAttr result will be cached and reused later during parallel
+				// hook calls
+				//
```
```suggestion
				// VersionedAttr result will be cached and reused later during parallel
				// hook calls.
```

sttts commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/policy/mutating/accessor.go:
```diff
@@ -0,0 +1,144 @@
+	}
+}
+
+func convertV1alpha1ResourceRulesToV1(mc *v1alpha1.MatchResources) *v1.MatchResources {
```
can't this be generate by conversion-gen?

jpbetz replied:
Because we're only using the function to implement the above `GetMatchResources()`, my understanding conversion-gen doesn't know we need the function.  If we could indicate somehow that we need it for a purpose not directly tied to conversion, it would generate useful functions for us.   I think this is OK.

jpbetz commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/cel/activation.go:
```diff
@@ -0,0 +1,162 @@
+
+// newActivation creates an activation for CEL admission plugins from the given request, admission chain and
+// variable binding information.
+func newActivation(ctx context.Context, versionedAttr *admission.VersionedAttributes, request *admissionv1.AdmissionRequest, inputs OptionalVariableBindings, namespace *v1.Namespace) (*evaluationActivation, error) {
```
This is pre-existing code extracted out of the the below ForInput function.

sttts replied:
`filter.ForInput` actually. We now have another `evaluator.ForInput` func.

jpbetz commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/cel/activation.go:
```diff
@@ -0,0 +1,162 @@
+
+// evaluateWithActivation evaluates a compiled CEL admission plugin expression using the provided activation and CEL
+// runtime cost budget.
+func evaluateWithActivation(ctx context.Context, activation *evaluationActivation, compilationResult CompilationResult, remainingBudget int64) (EvaluationResult, int64, error) {
```
This is also pre-existing code extracted out of the below ForInput function.

sttts replied:
Could be a method `Evaluate` on evaluationActivation.

jpbetz replied:
Nice. Done.

MariamFahmy98 commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/policy/mutating/plugin.go:
```diff
@@ -0,0 +1,151 @@
+
+// NewPlugin returns a generic admission webhook plugin.
+func NewPlugin(_ io.Reader) *Plugin {
+	// There is no request body to mutate for DELETE and CONNECT, so this plugin never handles those operations.
+	handler := admission.NewHandler(admission.Create, admission.Update, admission.Connect)
```
I think `admission.Connect` should be removed since there will be no mutations for CONNECT operations.

jpbetz replied:
This was previously discussed: https://github.com/kubernetes/kubernetes/pull/127134#discussion_r1750791786

We do support mutations for CONNECT for mutating webhooks as well.

MariamFahmy98 replied:
I think the comment needs to be updated.

sttts commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/cel/compile.go:
```diff
@@ -227,45 +229,70 @@ func mustBuildEnvs(baseEnv *environment.EnvSet) variableDeclEnvs {
+		},
+	)
+	if err != nil {
+		panic(fmt.Sprintf("environment misconfigured: %v", err))
```
nit: as this is not a `must*` function anymore, maybe return an error and panic from the caller.

sttts commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/cel/interface.go:
```diff
@@ -95,3 +98,23 @@ type Filter interface {
+// It expects the inbound object to already have been converted to the version expected
+// by the underlying CEL code (which is indicated by the match criteria of a policy definition).
+// versionedParams may be nil.
+type Evaluator interface {
```
it's unclear to me how this differs from `CompilationResult`. One of these is likely badly named. `CompilationResult`'s go doc talks about validation. Is that the difference?

sttts replied:
Next question: `Filter` is used for validation, `Evaluator` is used for mutation. Neither is really apparent from the name, and both have **exactly** the same go doc. These two concepts need better names and better docs. How do they differ? Is a `Filter` actually a `Validator`? What does it filter after all?

sttts replied:
Or `ValidatingEvaluator` and `MutatingEvaluator`.

jpbetz replied:
I like `MutatingEvaluator`.  I'd like to use `ConditionEvaluator` instead of `ValidatingEvaluator` since it is used for both validations and match conditions. I'll clarify that detail in the godoc.

sttts commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/cel/interface.go:
```diff
@@ -95,3 +98,23 @@ type Filter interface {
+type Evaluator interface {
+	// ForInput converts compiled CEL-typed values into a CEL-typed value to be patched.
+	// runtimeCELCostBudget was added for testing purpose only. Callers should always use const RuntimeCELCostBudget from k8s.io/apiserver/pkg/apis/cel/config.go as input.
+	// If cost budget is calculated, the filter should return the remaining budget.
```
what is a filter? Outdated comment?

sttts commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/cel/activation.go:
```diff
@@ -0,0 +1,162 @@
+// variable binding information.
+func newActivation(ctx context.Context, versionedAttr *admission.VersionedAttributes, request *admissionv1.AdmissionRequest, inputs OptionalVariableBindings, namespace *v1.Namespace) (*evaluationActivation, error) {
+	// if this activation supports composition, we will need the compositionCtx. It may be nil.
+	compositionCtx, _ := ctx.(CompositionContext)
```
this is odd. Why not just make the argument a `CompositionContext`?

sttts commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/cel/activation.go:
```diff
@@ -0,0 +1,162 @@
+	// if this activation supports composition, we will need the compositionCtx. It may be nil.
+	compositionCtx, _ := ctx.(CompositionContext)
+
+	var err error
+
```
```suggestion
```

sttts commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/cel/activation.go:
```diff
@@ -0,0 +1,162 @@
+// runtime cost budget.
+func evaluateWithActivation(ctx context.Context, activation *evaluationActivation, compilationResult CompilationResult, remainingBudget int64) (EvaluationResult, int64, error) {
+	// if this evaluation supports composition, we will need the compositionCtx. It may be nil.
+	compositionCtx, _ := ctx.(CompositionContext)
```
same comment as above: why not make it type safe by changing the function type?

jpbetz replied:
I've moved this up on level, the pre-existing code does pass this information around in a context so I do end up needing to do this somewhere, but moving it up the stack one level does simplify it a bit.

sttts commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/policy/generic/plugin.go:
```diff
@@ -181,7 +182,14 @@ func (c *Plugin[H]) ValidateInitialization() error {
+	}()
+
+	go func() {
+		err := c.dispatcher.Run(pluginContext)
```
you write elsewhere that `Run` does not block. Then we don't need this go routine.

jpbetz replied:
I've removed this go routine.  I can't find any reason to keep it.

cc @alexzielenski

sttts commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/policy/generic/policy_dispatcher.go:
```diff
@@ -148,17 +151,15 @@ func (d *policyDispatcher[P, B, E]) Dispatch(ctx context.Context, a admission.At
 				continue
 			} else if !matches {
 				continue
+			} else if _, err = versionedAttrAccessor.VersionedAttribute(matchGVK); err != nil {
+				// VersionedAttr result will be cached and reused later during parallel
+				// hook calls.
+				addConfigError(err, policyAccessor, nil)
+				continue
```
Took me a minute to understand that this has nothing to do with the upper binding match call. It's just the continuation of the case where the binding matches. Hence:
```suggestion
			}
			
			// here the binding matches
			if _, err = versionedAttrAccessor.VersionedAttribute(matchGVK); err != nil {
				// VersionedAttr result will be cached and reused later during parallel
				// hook calls.
				addConfigError(err, policyAccessor, nil)
				continue
```

sttts commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/policy/generic/policy_dispatcher.go:
```diff
@@ -194,23 +188,67 @@ func (d *policyDispatcher[P, B, E]) Dispatch(ctx context.Context, a admission.At
-		// no matching hooks
-		return nil
+	if len(filteredErrors) > 0 {
+		var err *apierrors.StatusError
+		if !errors.As(admission.NewForbidden(a, fmt.Errorf("admission request denied by policy")), &err) {
+			return apierrors.NewInternalError(fmt.Errorf("failed to create status error"))
+		}
+		err.ErrStatus.Message = ""
```
What is this logic about? Reading `NewForbidden` it either returns `err` if it is already a forbidden error (it isn't here), or it returns `apierrors.NewInternalError(...)` or `apierrors.NewForbidden(...)`. In either case it is a status error. If we don't want to change `admission.NewForbidden`'s signature (maybe the best direction, but it has ripples), then:
```suggestion
		err, ok := admission.NewForbidden(...)
		if !ok {
			return apierrors.NewInternalError(fmt.Errorf("failed to create status error"))
		}
		err.ErrStatus.Message = ""
```
A lot easier to read.

jpbetz replied:
The linter didn't like:

```go
err, ok := admission.NewForbidden(a, fmt.Errorf("admission request denied by policy")).(*apierrors.StatusError)
if !ok {
	return apierrors.NewInternalError(fmt.Errorf("failed to create status error"))
}
```

```
ERROR: staging/src/k8s.io/apiserver/pkg/admission/plugin/policy/generic/policy_dispatcher.go:229:14: type assertion on error will fail on wrapped errors. Use errors.As to check for specific errors (errorlint)
ERROR: 		err, ok := admission.NewForbidden(a, fmt.Errorf("admission request denied by policy")).(*apierrors.StatusError)
ERROR: 		           ^

Please review the above warnings. You can test via "./hack/verify-golangci-lint.sh -r 43ad87b85a39344db31323b59c2eab4adc73b66e -n"
```

So I've restructured this a bit and added comments to make it clearer.

sttts commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/policy/generic/policy_dispatcher.go:
```diff
@@ -194,23 +188,67 @@ func (d *policyDispatcher[P, B, E]) Dispatch(ctx context.Context, a admission.At
-				return apierrors.NewInternalError(err)
-			}
+	if len(relevantHooks) > 0 {
+		extraPolicyErrors, statusError := d.delegate(ctx, a, o, versionedAttrAccessor, relevantHooks)
```
have not checked: this change makes `d.delegate` not see the hooks that have `hook.Error` set. What is the impact of that change?

jpbetz replied:
@alexzielenski do you remember?

alexzielenski replied:
I think @sttts pointed out in another comment that the policy dispatcher was not yet being used so there is no impact

sttts commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/policy/generic/policy_dispatcher.go:
```diff
@@ -76,7 +72,7 @@ type PolicyInvocation[P runtime.Object, B runtime.Object, E Evaluator] struct {
 // The delegate provides the "validation" or "mutation" aspect of dispatcher functionality
 // (in contrast to generic.PolicyDispatcher which only selects active policies and params)
-type dispatcherDelegate[P, B runtime.Object, E Evaluator] func(ctx context.Context, a admission.Attributes, o admission.ObjectInterfaces, versionedAttributes webhookgeneric.VersionedAttributeAccessor, invocations []PolicyInvocation[P, B, E]) error
+type dispatcherDelegate[P, B runtime.Object, E Evaluator] func(ctx context.Context, a admission.Attributes, o admission.ObjectInterfaces, versionedAttributes webhookgeneric.VersionedAttributeAccessor, invocations []PolicyInvocation[P, B, E]) ([]PolicyError, *apierrors.StatusError)
```
While trying to understand the dispatcher concept I noticed that it is only used in mutating admission, but was preexisting before the PR? `NewPolicyDispatcher` not called from the validating code. Why is that?

jpbetz replied:
Yes, it looks like validating admission retains it's own unshared dispatcher.  This might to insulate ValidatingAdmissionPolicy from the changes in this PR?

cc @alexzielenski for any additional context/history

alexzielenski replied:
The VAP dispatcher embeds the generic dispatcher. There are separate dispatchers because for validating vs mutating case the actual evaluation of what the hook does is different. The validating dispatcher evaluates the policies against the request and returns an error without modifying anything. The mutating dispatcher will change the object according to the policy. But both have shared code in `generic.Dispatcher` used for corralling all the policies, bindings, params, etc.

jpbetz replied:
Thanks @alexzielenski. We can follow up on the code sharing in follow up PRs.

sttts commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/cel/evaluator.go:
```diff
@@ -0,0 +1,75 @@
+	}
+	return evaluation, remainingBudget, nil
+
+}
```
looking at this function and on `filter`, hot take: isn't `evaluator` just a single expression `filter`. Trying to spot the difference. Is there any?

jpbetz replied:
Yes, and filter also has the expectation that each expression evaluates to a bool.  I've restructured this based on above comments.  Let me know if it makes sense. Happy to refine this further.

sttts commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/cel/compile.go:
```diff
@@ -227,45 +229,70 @@ func mustBuildEnvs(baseEnv *environment.EnvSet) variableDeclEnvs {
+		if err != nil {
+			panic(fmt.Sprintf("environment misconfigured: %v", err))
+		}
+	}
```
for fellow reviewers: these lines are what is new. The rest is moved around code.

sttts commented on test/integration/apiserver/cel/mutatingadmissionpolicy_test.go:
```diff
@@ -0,0 +1,1418 @@
+	dynClient, err := dynamic.NewForConfig(server.ClientConfig)
+	require.NoError(t, err)
+
+	sch := runtime.NewScheme()
```
not seeing this used.

jpbetz replied:
Removed

sttts commented on test/integration/apiserver/cel/mutatingadmissionpolicy_test.go:
```diff
@@ -0,0 +1,1418 @@
+			dynClient, err := dynamic.NewForConfig(server.ClientConfig)
+			require.NoError(t, err)
+
+			sch := runtime.NewScheme()
```
not seeing this used.

jpbetz replied:
Removed

sttts commented on test/integration/apiserver/cel/mutatingadmissionpolicy_test.go:
```diff
@@ -0,0 +1,1418 @@
+			require.NoError(t, err)
+			defer server.TearDownFn()
+
+			client, err := clientset.NewForConfig(server.ClientConfig)
```
nit: elsewhere clientset is imported as kubernetes in this PR

sttts commented on test/integration/apiserver/cel/mutatingadmissionpolicy_test.go:
```diff
@@ -0,0 +1,1418 @@
+
+			if tc.initialObject != nil {
+				initClient := clientForType(t, tc.initialObject, tc.requestResource, dynClient)
+				_, err = initClient.Create(context.TODO(), toUnstructured(t, tc.initialObject), metav1.CreateOptions{})
```
Elsewhere we do
```
ctx, cancel := context.WithCancel(context.Background())
t.Cleanup(cancel)
```
and avoid `context.TODO()`.

sttts commented on staging/src/k8s.io/apiserver/pkg/cel/mutation/typeresolver.go:
```diff
@@ -0,0 +1,47 @@
+	if name == JSONPatchTypeName {
+		return &JSONPatchType{}, true
+	}
+	if strings.HasPrefix(name, ObjectTypeName) {
```
`if name == ObjectTypeName || strings.HasPrefix(name, ObjectTypeName + ".") {`?

sttts commented on staging/src/k8s.io/apiserver/pkg/cel/mutation/jsonpatch.go:
```diff
@@ -0,0 +1,183 @@
+	"reflect"
+)
+
+var jsonPatchType = types.NewObjectType("JSONPatch")
```
JSONPatchTypeName also exists in this package

sttts commented on staging/src/k8s.io/apiserver/pkg/cel/mutation/jsonpatch.go:
```diff
@@ -0,0 +1,183 @@
+	if typeDesc == reflect.TypeOf(&JSONPatchVal{}) {
+		return p, nil
+	}
+	return nil, fmt.Errorf("cannot convert to native type: %v", typeDesc)
```
not print `%T`? Or will `%v` with `typeDesc` give some nice cel string representation?

jpbetz replied:
I think this is OK. This funnels into `reflect.Type`s `String()` function which I think is very similar to calling `%T` on the reflected value.

sttts commented on staging/src/k8s.io/apiserver/pkg/cel/mutation/jsonpatch.go:
```diff
@@ -0,0 +1,183 @@
+	if name, ok := field.Value().(string); ok {
+		switch name {
+		case "op":
+			if len(p.Op) > 0 {
```
`return types.Bool(len(p.Op) > 0)`. Same everywhere else.

jpbetz replied:
That is much cleaner. Thx.

sttts commented on staging/src/k8s.io/apiserver/pkg/cel/mutation/dynamic/objects.go:
```diff
@@ -0,0 +1,245 @@
+	"github.com/google/cel-go/common/types/traits"
+	"google.golang.org/protobuf/types/known/structpb"
+	"reflect"
+	"strings"
```
order

sttts replied:
also in a couple of other files here

jpbetz replied:
Oops, I'll do a quick sweep.

sttts commented on staging/src/k8s.io/apiserver/pkg/cel/mutation/dynamic/objects.go:
```diff
@@ -0,0 +1,245 @@
+	objectType *types.Type
+}
+
+func (r *ObjectType) HasTrait(trait int) bool {
```
nit: `r` like ObjectType? Or like ref?

jpbetz replied:
Renamed to `o`.  Can't remember why `r` ended up being used there.

sttts commented on staging/src/k8s.io/apiserver/pkg/cel/mutation/dynamic/objects.go:
```diff
@@ -0,0 +1,245 @@
+
+// Equal returns true if the `other` value has the same type and content as the implementing struct.
+func (v *ObjectVal) Equal(other ref.Val) ref.Val {
+	if rhs, ok := other.(*ObjectVal); ok {
```
why not check the `objectType` first?

jpbetz replied:
Good idea. Added.

cici37 commented on staging/src/k8s.io/apiserver/pkg/cel/mutation/dynamic/objects.go:
```diff
@@ -0,0 +1,245 @@
+
+}
+
+func typeCheck(v ref.Val, typeNamePath []string) []error {
```
No test added for the typeCheck?

jpbetz replied:
I've added tests for CheckTypeNamesMatchFieldPathNames which cover this.

jpbetz commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/cel/activation.go:
```diff
@@ -0,0 +1,162 @@
+
+	oldObjectVal, err := objectToResolveVal(versionedAttr.VersionedOldObject)
+	if err != nil {
+		return nil, err
```
Ah, this is pre-existing code I moved. I'll add contexts.

cici37 commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/policy/mutating/patch/json_patch.go:
```diff
@@ -0,0 +1,172 @@
+				// Before beta, attaching full type information both to Object initializers and
+				// the "object" and "oldObject" variables is needed. This will allow CEL to
+				// perform comprehensive runtime type checking.
+				err := objVal.CheckTypeNamesMatchFieldPathNames()
```
Worth adding tests to show the current type check support and limit?

jpbetz replied:
Added.

jpbetz commented on test/integration/apiserver/cel/mutatingadmissionpolicy_test.go:
```diff
@@ -0,0 +1,1418 @@
+	// Run all tests in a shared apiserver
+	featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, genericfeatures.MutatingAdmissionPolicy, true)
+	server, err := apiservertesting.StartTestServer(t, nil, []string{
+		"--enable-admission-plugins", "MutatingAdmissionPolicy",
```
I don't think this is actually needed. I'll remove.

jpbetz commented on staging/src/k8s.io/apiserver/pkg/cel/mutation/typeresolver_test.go:
```diff
@@ -0,0 +1,259 @@
+			name:          "equality check",
+			expression:    "Object{spec: Object.spec{replicas: 3}} == Object{spec: Object.spec{replicas: 1 + 2}}",
+			expectedValue: true,
+		},
```
Added a test for this. Thanks for the fix.

jpbetz commented on staging/src/k8s.io/apiserver/pkg/cel/mutation/jsonpatch.go:
```diff
@@ -0,0 +1,183 @@
+}
+
+func (r *JSONPatchType) FieldNames() ([]string, bool) {
+	return []string{"op", "from", "patch", "value"}, true
```
Yes, that helps here. Thanks for catching that error also.

jpbetz commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/cel/activation.go:
```diff
@@ -0,0 +1,162 @@
+// variable binding information.
+func newActivation(ctx context.Context, versionedAttr *admission.VersionedAttributes, request *admissionv1.AdmissionRequest, inputs OptionalVariableBindings, namespace *v1.Namespace) (*evaluationActivation, error) {
+	// if this activation supports composition, we will need the compositionCtx. It may be nil.
+	compositionCtx, _ := ctx.(CompositionContext)
```
Yes, below in this function is an if-nil check.

jpbetz commented on staging/src/k8s.io/apiserver/pkg/cel/common/typeprovider_test.go:
```diff
@@ -0,0 +1,149 @@
+			name:          "empty",
+			expression:    "Test{}",
+			expectedValue: map[string]any{},
+		},
```
The test resolver registers for `Test`, not `Object`.  I've expanded the tests from the code I inherited here.

Note that `typeresolver_test.go` does test the `Object` resolver.

jpbetz replied:
Added more tests.  Note that "Test{}" here is the valid object.

jpbetz commented on staging/src/k8s.io/apiserver/pkg/cel/mutation/typeresolver_test.go:
```diff
@@ -0,0 +1,259 @@
+func mustCreateEnv(t testing.TB, envOptions ...cel.EnvOption) *cel.Env {
+	envSet, err := environment.MustBaseEnvSet(environment.DefaultCompatibilityVersion(), true).
+		Extend(environment.VersionedOptions{
+			IntroducedVersion: version.MajorMinor(1, 30),
```
Extend will raise a nil panic if IntroducedVersion is not provided on a VersionedOptions.  I'll set this to 1.0 and add a comment.

jpbetz commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/cel/compile.go:
```diff
@@ -227,45 +229,70 @@ func mustBuildEnvs(baseEnv *environment.EnvSet) variableDeclEnvs {
+	EnvOptions: []cel.EnvOption{
+		common.ResolverEnvOption(&mutation.DynamicTypeResolver{}),
+		library.JSONPatch(), // for jsonPatch.escape() function
+	},
```
This seemed lower risk.... If we do this, we could choose at a later time to share it with the base env, but if we add it now to the base environment we can never remove it.  I'm in favor of waiting until we have a good reason to add it to the base env.

jpbetz commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/policy/mutating/compilation_test.go:
```diff
@@ -0,0 +1,976 @@
+				"same":      "true",
+				"different": "false",
+			}}},
+		},
```
Done.

> Tests could be added in the compilation/evaluation level:
> 
> * jsonpatch escape

See the "JSONPatch key escaping" test.

> * variables used in applyconfiguration

See "apply configuration with variable"

> * current limited type checking support and a TODO

Good idea. Added.

> * different cel variable support(params, oldObject, namespace, request, authorizer, etc)

- params - Added
- oldObject - See "apply configuration with old object"
- namespace - Added
- request - See "request variable"
- authorizer - See "authorizer check"

> * cost limitation for both paths

Added for apply configurations. See "jsonPatch with excessive cost" for jsonpatch.

jpbetz commented on test/integration/apiserver/cel/mutatingadmissionpolicy_test.go:
I believe you'd written all of these.

> Worth adding the integration test for
> 
> * reinvocation

See "multiple policies requiring reinvocation" and "multiple params causing reinvocation"

> * mutation ordering

See "multiple params causing reinvocation". This one checks ordering in the presence of reinvocation.

> * variables used in expression

See "policy with variables"

cici37 commented on pkg/apis/admissionregistration/types.go:
```diff
@@ -1163,3 +1164,327 @@ type MatchCondition struct {
+	//
+	// CEL expressions have access to the contents of the API request, organized into CEL variables as well as some other useful variables:
+	//
+	// - 'object' - The object from the incoming request. The value is null for DELETE requests.
```
I think the object refers to the object after previous mutation and before current mutation which could be different from incoming request. Maybe clarifying in the doc to avoid confusion?

jpbetz replied:
This behavior is identical to a webhook `AdmissionReview`s `object` field (https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/#request).  But we will make sure to document it carefully.

cici37 reviewed (commented):
Notes to myself and other reviewers so far:
1. [Expected]For applyconfiguration patch, unset value using optional.none() works for primitive type but not supported for struct(like delete a specific item from map/list).
2. [Expected]Ordering is not guaranteed which means if having multiple MAP to reset the same field with different value, the final state of that field is not deterministic.
    The mutations defined inside one single MAP would be evaluated in order and it could be leveraged to mutate the object in sequence.
3. The field value defaulting happens before mutation admission which means user has to explicitly state all required field value when needed(e.g. when insert new container).
~~EDIT: remove these lines based on following comments~~
4. cel variables(object, oldObject, request, namespaceObject etc), variables field, matchConditions, etc fields work as expected.
5. Note that the `object` refers to the object after previous mutation and before current mutation. It needs more clear doc as it might confuse user. 
6. Currently type checking and schema conversion err will only show up in runtime which raised the bar for policy author.

Overall it looks great! Cann't wait for it to be merged :)

jpbetz commented:
> 4. Construct a struct as an item inside the map/list is not supported currently?

This is supported and tested. See https://github.com/jpbetz/kubernetes/blob/b8c1f3fb7d0c44ae885c9709ba8263c895db334d/staging/src/k8s.io/apiserver/pkg/admission/plugin/policy/mutating/compilation_test.go#L538-L540

> 5. Modify list items/associated list remain unsupported for ApplyConfiguration(might remain unsupported).

This is supported. See above mentioned test. The key fields of the list item provide the association.

cici37 commented:
> > 4. Construct a struct as an item inside the map/list is not supported currently?
> 
> This is supported and tested. See https://github.com/jpbetz/kubernetes/blob/b8c1f3fb7d0c44ae885c9709ba8263c895db334d/staging/src/k8s.io/apiserver/pkg/admission/plugin/policy/mutating/compilation_test.go#L538-L540
> 

For example, to set `imagePullPolicy` to `always` to all containers in the pod:
```
expression: 'Object{
					spec: Object.spec{
						containers: [object.spec.containers.map(c, 
						Object.spec.containers{
                                                       name: c.name,
                                                       imagePullPolicy: "Always"
                                                   })]}}'
```
returns err: `error applying patch: failed to convert patch object to typed object: .spec.containers: element 0: associative list with keys may not have non-map elements`. Is the expression not in expected form?

jpbetz commented:
> returns err: `error applying patch: failed to convert patch object to typed object: .spec.containers: element 0: associative list with keys may not have non-map elements`. Is the expression not in expected form?

`containers: [object.spec.containers.map(c, ...)]` attempts to set containers to a list of lists.  This should instead be `containers: object.spec.containers.map(c, ...)`.  This will be improved when we do a full type check against actual runtime types, which I'm adding to the beta graduation criteria.

sttts commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/cel/condition.go:
```diff
@@ -0,0 +1,216 @@
+		}
+	}
+	return compilationErrors
+}
```
this file is just a rename, which github does not understand?

sttts replied:
oh plus use of activation I guess.

jpbetz replied:
Correct

jpbetz replied:
I've squashed. This now shows up cleanly as a rename.

sttts commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/cel/interface.go:
```diff
@@ -82,16 +85,38 @@ type OptionalVariableBindings struct {
+// It expects the inbound object to already have been converted to the version expected
+// by the underlying CEL code (which is indicated by the match criteria of a policy definition).
+// versionedParams may be nil.
+type MutatingEvaluator interface {
```
a lot better with the rename.

sttts commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/policy/mutating/dispatcher.go:
```diff
@@ -0,0 +1,284 @@
+				// This should never happen. It occurs if there is a programming
+				// error causing the Param not to be a valid object.
+				return nil, k8serrors.NewInternalError(err)
+			}
```
This looks fishy. See 20 lines further down where the loop continues on (non status) error. The other case is when the err is set by `keyFor` above.

sttts replied:
And for `keyFor` better do error handling near to it, very explicitly, not 10 lines further down in a loop.

jpbetz replied:
Oh wow, very good catch. I'll use a uniquely named error variable for this special case check.

sttts commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/policy/mutating/dispatcher.go:
```diff
@@ -0,0 +1,284 @@
+			}
+		}
+		if !apiequality.Semantic.DeepEqual(objectBeforeMutations, versionedAttr.VersionedObject) {
+			// Patch had changed the object. Prepare to reinvoke all previous webhooks that are eligible for re-invocation.
```
Copied from webhook mutation. I guess neither the original nor this copy here are correct anymore.

jpbetz replied:
Fixed

sttts commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/policy/mutating/patch/json_patch.go:
```diff
@@ -0,0 +1,173 @@
+			// If a json patch fails a test operation, the patch must not be applied
+			return r.VersionedAttributes.VersionedObject, nil
+		}
+		return nil, fmt.Errorf("JSONPath: %w", err)
```
JSON path? Patch? How does this error look like?

jpbetz replied:
Ooh, good catch. This should be JSONPatch

sttts commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/policy/mutating/patch/json_patch.go:
```diff
@@ -0,0 +1,173 @@
+
+import (
+	"context"
+	json2 "encoding/json"
```
nit: gojson

sttts commented on staging/src/k8s.io/pod-security-admission/admission/admission.go:
```diff
@@ -191,7 +191,7 @@ func (a *Admission) ValidateConfiguration() error {
 		return fmt.Errorf("PodSpecExtractor required")
 	}
 	if a.Evaluator == nil {
-		return fmt.Errorf("Evaluator required")
```
replace error

sttts commented on staging/src/k8s.io/apiserver/pkg/cel/mutation/typeresolver_test.go:
```diff
@@ -0,0 +1,279 @@
+	"github.com/google/cel-go/common/types/ref"
+	"reflect"
+	"strings"
+	"testing"
```
order

jpbetz committed:
Move caching authorizer to shared location to be used by mutating and validating policy

jpbetz committed:
Clean up Object initialization support

The initial work of this had been merged before
this PR but was not yet in use. This simplifies
the implementation and adds some basic type
sanity checking.

Co-authored-by: Jiahui Feng <jhf@google.com>

jpbetz committed:
Add feature gate

Co-authored-by: cici37 <cicih@google.com>

jpbetz committed:
Add MutatingAdmissionPolicy API

This is closely aligned with ValidatingAdmissionPolicy
except that instead of validations that can fail with
messages, there are mutations, which can be defined
either with as an ApplyConfiguration or JSONPatch.

Co-authored-by: cici37 <cicih@google.com>

jpbetz committed:
generate code

jpbetz committed:
Add MutatingAdmissionPolicy plugin to admission chain

This expands the generic plugin support to both validating and mutating policies.  It also adds the
mutating policy admission plugin using the generics plugin support.

This also implements both ApplyConfiguration and JSONPatch support.

Co-authored-by: Alexander Zielensk <alexzielenski@gmail.com>

jpbetz committed:
Add jsonpatch.escapeKey CEL function

jpbetz committed:
Add integration tests

Co-authored-by: cici37 <cicih@google.com>
Co-authored-by: Alexander Zielensk <alexzielenski@gmail.com>

jpbetz committed:
Update OpenAPI and fix openAPI tests to handle unexported jsonreferences

Co-authored-by: Alexander Zielensk <alexzielenski@gmail.com>

jpbetz committed:
Add MutatingAdmissionPolicy to samples

jpbetz committed:
Improve error messaging for validating admission policy authz

jpbetz committed:
Add mutation support into CompositedCompiler and reorganize for clarity

jpbetz committed:
Fix comment in mutating webhook dispatcher

jpbetz commented:
Thanks @sttts!   I've merged in the last round of feedback and then rebased and squashed all feedback commits into the main commits.

jpbetz committed:
Add authz test for deny when relevant fields are not changed

jpbetz committed:
Rename dispatcher Run to Start to match naming conventions

sttts commented on pkg/registry/admissionregistration/validatingadmissionpolicy/storage/storage_test.go:
```diff
@@ -225,3 +226,11 @@ func TestCategories(t *testing.T) {
 	registrytest.AssertCategories(t, storage, expected)
 }
+
+var replicaLimitsResolver resolver.ResourceResolverFunc = func(gvk schema.GroupVersionKind) (schema.GroupVersionResource, error) {
```
nit: `func replicaLimitsResolver(gvk schema.GroupVersionKind) (schema.GroupVersionResource, error)`

jpbetz replied:
But I can do `resolver.ResourceResolverFunc(replicaLimitsResolver)` at the call site, so I'll do that.

jpbetz committed:
Reorganize and expand unit test coverage

Also apply reviewer feedback

sttts commented on staging/src/k8s.io/apiserver/pkg/cel/mutation/jsonpatch.go:
```diff
@@ -0,0 +1,178 @@
+	} else if typeValue == types.TypeType {
+		return types.NewTypeTypeWithParam(jsonPatchType)
+	}
+	return types.NewErr("Unsupported type: %s", typeValue.TypeName())
```
nit: lower case

sttts commented on staging/src/k8s.io/apiserver/pkg/cel/mutation/jsonpatch.go:
```diff
@@ -0,0 +1,178 @@
+
+func (p *JSONPatchVal) Equal(other ref.Val) ref.Val {
+	if o, ok := other.(*JSONPatchVal); ok && p != nil && o != nil {
+		if *p == *o {
```
why is this the right equality? Can the value be any JSON value including maps and arrays?

jpbetz replied:
Good catch. This does not look right. I've fixed with a sane equality implementation.

sttts commented on staging/src/k8s.io/apiserver/pkg/cel/mutation/dynamic/objects.go:
```diff
@@ -0,0 +1,249 @@
+		for k, v := range mapOfVal {
+			stringKey, ok := k.Value().(string)
+			if !ok {
+				return nil, fmt.Errorf("map key %q is of type %t, not string", k, k)
```
```suggestion
				return nil, fmt.Errorf("map key %q is of type %T, not string", k, k)
```

sttts commented on staging/src/k8s.io/apiserver/pkg/cel/mutation/dynamic/objects.go:
```diff
@@ -0,0 +1,249 @@
+	// unstructured maps, as seen in annotations
+	// map keys must be strings
+	if mapOfVal, ok := value.Value().(map[ref.Val]ref.Val); ok {
+		result := make(map[string]any)
```
this does not distinguish between nil and empty. Are there cases where we should? I guess no, but how can we know?

jpbetz replied:
> this does not distinguish between nil and empty. Are there cases where we should? I guess no, but how can we know?

Good question. It looks like this is processing `ref.Val`, so we're guaranteed to be processing only valid CEL values, which simplifies things a bit.  CEL does have null, but it is not used as a zero value for a maps or slices, so we're safe here.

sttts commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/policy/validating/typechecking.go:
```diff
@@ -436,7 +436,7 @@ func buildEnvSet(hasParams bool, hasAuthorizer bool, types typeOverwrite) (*envi
 }
 
-// createVariableOpts creates a slice of EnvOption
+// createVariableOpts creates a slice of ResolverEnvOption
```
replace mistake?

jpbetz replied:
Thanks, yes. Removing.

sttts commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/policy/mutating/reinvocationcontext.go:
```diff
@@ -0,0 +1,76 @@
+	rc.lastPolicyOutput = object.DeepCopyObject()
+}
+
+func (rc *policyReinvokeContext) AddReinvocablePolicyToPreviouslyInvoked(policy key) {
```
Seeing `// if preconditions are not met, then skip mutations` in the dispatcher, skipping the call of this method. I.e. one policy mutating in a way that another policy applies does not cause the later to be called. Is that intentional?

jpbetz replied:
> what's the semantics here? If a policy is not applied due to some match conditions, but for an reinvocation a match conditions would actually match, would that policy show up here or get ignored?

No, this is not supported for match conditions (for webhooks or for policies).  I had not realize this is a gap until now! I've added a testing showing the gap.  I'd be happy to look into it.  OK if we follow up post-freeze, make this beta graduation criteria?

jpbetz replied:
> one policy mutating in a way that another policy applies does not cause the later to be called. Is that intentional?

I've also added a test showing that if the first policy adds a label that the next policy has a match condition for, that the second policy is evaluated.  But my above comment applies to the general case.

sttts replied:
Making it a beta graduation criteria sounds fine. Am not even sure it is that clear which behaviour we want. Reinvocation is always kind of best effort. Sometimes it might need more than one reinvocation to converge, which we obviously never do. So deciding and documenting is also an option for the behaviour here.

sttts commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/policy/mutating/dispatcher.go:
```diff
@@ -0,0 +1,284 @@
+	res := &dispatcher{
+		matcher: m,
+		authz:   a,
+		//!TODO: pass in static type converter to reduce network calls
```
It's about the openapi 30s loop, right? It's not a network call in a critical path. Maybe be more specific here. In 12 months we will have forgotten what is meant here.

sttts replied:
Nvmd. Not 30s. It's on demand, but cached. Still not on the critical path.

jpbetz replied:
Looks like is copy-paste originating from the webhook reinvocation code. Removing.

sttts commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/policy/mutating/dispatcher.go:
```diff
@@ -0,0 +1,284 @@
+	}
+
+	return policyErrors, nil
+}
```
am I right that all this logic here is untested? At least no unit tests, e.g. about reinvocations.

jpbetz replied:
This has some coverage in plugin_test.go, but I'll add dedicated unit testing.

jpbetz replied:
Added unit test for both the dispatcher (including reinvocation test cases) and for the reinvocation context file.

sttts commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/policy/mutating/patch/typeconverter.go:
```diff
@@ -0,0 +1,187 @@
+// type converter for our openapi models. It should be connfigured to use a
+// static type converter for natively typed schemas, and fetches the schema
+// for CRDs/other over the network on demand (trying to reduce network calls where necessary)
+type typeConverterManager struct {
```
this is untested.

jpbetz replied:
Added a test that verifies the core behavior.

sttts commented on staging/src/k8s.io/apiserver/pkg/cel/mutation/jsonpatch.go:
```diff
@@ -0,0 +1,184 @@
+		}
+		if p.Val != nil {
+			return p.Val.Equal(o.Val)
+		}
```
better

sttts commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/policy/generic/policy_dispatcher.go:
```diff
@@ -352,3 +400,18 @@ func (v *versionedAttributeAccessor) VersionedAttribute(gvk schema.GroupVersionK
+		return fmt.Sprintf("policy '%s' with binding '%s' denied request: %s", c.Policy.GetName(), c.Binding.GetName(), c.Message.Error())
+	}
+
+	return fmt.Sprintf("policy '%s' denied request: %s", c.Policy.GetName(), c.Message.Error())
```
nit: %q

sttts commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/policy/mutating/dispatcher.go:
```diff
@@ -0,0 +1,284 @@
+		if invocation.Evaluator.Matcher != nil {
+			matchResults := invocation.Evaluator.Matcher.Match(ctx, versionedAttr, invocation.Param, authz)
+			if matchResults.Error != nil {
+				addConfigError(matchResults.Error, invocation, metav1.StatusReasonInvalid)
```
there are no cases where `Error` is set and `Matches` is true. A `continue` would simplify logic a little.

sttts commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/policy/mutating/patch/interface.go:
```diff
@@ -0,0 +1,43 @@
+
+// Patcher provides a patch function to perform a mutation to an object in the admission chain.
+type Patcher interface {
+	Patch(ctx context.Context, request Request, runtimeCELCostBudget int64) (runtime.Object, error)
```
we require that this returns a different `runtime.Object`, i.e. patch is NOT in-place (compare how we set `objectBeforeMutations`). This should be clearly specified.

jpbetz replied:
Good call. I've added godoc to spell it out.

sttts commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/policy/mutating/patch/smd.go:
```diff
@@ -0,0 +1,194 @@
+}
+
+// validatePatch searches an apply configuration for any arrays, maps or structs elements that are atomic and returns
+// an error if any are found.
```
why is this? Why are those rejected?

jpbetz replied:
@deads2k and I agreed to disallow this because it can result int accidental deletion of data when the user intended to modify part of an atomic but didn't realize that if they do not include the other fields in the atomic, those other fields will be removed.  I'll document.

sttts commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/policy/mutating/plugin_test.go:
```diff
@@ -0,0 +1,292 @@
+						ApplyConfiguration: &v1alpha1.ApplyConfiguration{
+							Expression: "ignored, but required",
+						},
+						PatchType: v1alpha1.PatchTypeApplyConfiguration,
```
Missing tests for `PatchTypeJSONPatch`.

jpbetz replied:
Added

sttts commented:
/lgtm
/approve

/assign @deads2k 
for final API review and approval.

deads2k commented:
/approve

jpbetz commented:
/assign @seans3 or @soltysh would either of you be willing to approve the printer changes here?

ardaguclu commented:
/lgtm

ardaguclu commented:
/approve

aaron-prindle commented on pkg/features/versioned_kube_features.go:
```diff
@@ -266,7 +266,7 @@ var defaultVersionedKubernetesFeatureGates = map[featuregate.Feature]featuregate
 
 	genericfeatures.MutatingAdmissionPolicy: {
-		{Version: version.MustParse("1.30"), Default: false, PreRelease: featuregate.Alpha},
+		{Version: version.MustParse("1.32"), Default: false, PreRelease: featuregate.Alpha},
```
Discussed offline, the Compatility versions feature gate integration test that verifies gates are correct for n-1..3 versions (eg: binary @ 1.32 using --emulated-version=1.31 is in compliance with the feature gates from 1.31) already will work for this change as it is an alpha API and alpha APIs are allowed to be removed w.r.t to Compatibility versions
